demos/text_detection_demo/cpp_gapi/src/custom_kernels.cpp

```diff
@@ -0,0 +1,408 @@
+        groupMask[root1] = root2;
+    }
+}
+
+cv::Mat getAll(const std::vector<cv::Point>& points, const int w, const int h,
+               std::unordered_map<int, int>& groupMask) {
+    std::unordered_map<int, int> rootMap;
+    cv::Mat mask(h, w, CV_32S, cv::Scalar(0));
+    for (const auto& point : points) {
+        int pointRoot = findRoot(point.x + point.y * w, groupMask);
+        if (rootMap.find(pointRoot) == rootMap.end()) {
+            rootMap.emplace(pointRoot, static_cast<int>(rootMap.size() + 1));
+        }
+        mask.at<int>(point.x + point.y * w) = rootMap[pointRoot];
+    }
+    return mask;
+}
+
+cv::Mat decodeImageByJoin(const std::vector<float>& segmData,
+                          const std::vector<int>&   segmDataShape,
+                          const std::vector<float>& linkData,
+                          const std::vector<int>&   linkDataShape,
+                          float segmConfThreshold,
+                          float linkConfThreshold) {
+    const int h = segmDataShape[1];
+    const int w = segmDataShape[2];
+    std::vector<uchar> pixelMask(h * w, 0);
+    std::unordered_map<int, int> groupMask;
+    std::vector<cv::Point> points;
+    for (size_t i = 0; i < pixelMask.size(); i++) {
+        pixelMask[i] = segmData[i] >= segmConfThreshold;
+        if (pixelMask[i]) {
+            points.emplace_back(i % w, i / w);
+            groupMask[i] = -1;
+        }
+    }
+    std::vector<uchar> linkMask(linkData.size(), 0);
+    for (size_t i = 0; i < linkMask.size(); i++) {
+        linkMask[i] = linkData[i] >= linkConfThreshold;
+    }
+    size_t neighbours = size_t(linkDataShape[3]);
+    for (const auto& point : points) {
+        size_t neighbour = 0;
+        for (int ny = point.y - 1; ny <= point.y + 1; ny++) {
+            for (int nx = point.x - 1; nx <= point.x + 1; nx++) {
+                if (nx == point.x && ny == point.y)
+                    continue;
+                if (nx >= 0 && nx < w && ny >= 0 && ny < h) {
+                    uchar pixelValue = pixelMask[size_t(ny) * size_t(w) + size_t(nx)];
+                    uchar linkValue = linkMask[(size_t(point.y) * size_t(w) + size_t(point.x))
+                                                    *neighbours + neighbour];
+                    if (pixelValue && linkValue) {
+                        join(point.x + point.y * w, nx + ny * w, groupMask);
+                    }
+                }
+                neighbour++;
+            }
+        }
+    }
+    return getAll(points, w, h, groupMask);
+}
+
+void processData(const float* ptr, const size_t dataTotal, const cv::MatSize& dataShape,
+                 std::vector<float>& out, std::vector<int>& newShape) {
+    std::vector<float> inData(ptr, ptr + dataTotal);
+    inData = transpose4d(inData, dimsToShape(dataShape), {0, 2, 3, 1});
+    softmax(inData);
+    out = sliceAndGetSecondChannel(inData);
+    newShape = { dataShape[0], dataShape[2], dataShape[3], dataShape[1] / 2 };
```

Comment:
Why do you cut channels and use NHWC layout for out shapes?

Reply:
Also I think N is extra here.
You use dataShape[2], dataShape[3] for segmNewShape and C for linkNewShape.

Reply:
`Why?` - I can't really give exhaustive answer. It's bound with math in deeper functions such as `transpose4d`.

I think you're right about N, thanks